File: next/dweb-browser/MicroService/src/Ipc/Message/IpcStreamData.cs
namespace DwebBrowser.MicroService.Message;

public class IpcStreamData : IpcMessage, IpcStream
{
    [JsonPropertyName("stream_id")]
    public string StreamId { get; set; }

    [JsonPropertyName("data")]
    public object _Data { get; set; }
    public object Data
    {
        get => _Data switch
        {
            JsonElement element => _Data = element.GetString()!, // JSON 模式下，只可能输出字符串格式，不可能是 byte[]
            /// TODO 未来支持 CBOR 的时候，这里可以直接读取出 byte[]
            _ => _Data,
        };
        set => _Data = value;
    }

    [JsonPropertyName("encoding")]
    public IPC_DATA_ENCODING Encoding { get; set; }

    [Obsolete("使用带参数的构造函数", true)]
    public IpcStreamData() : base(IPC_MESSAGE_TYPE.STREAM_DATA)
    {
        /// 给JSON反序列化用的空参数构造函数
    }
    public IpcStreamData(string stream_id, object data, IPC_DATA_ENCODING encoding) : base(IPC_MESSAGE_TYPE.STREAM_DATA)
    {
        StreamId = stream_id;
        Data = data;
        Encoding = encoding;
    }

    public static IpcStreamData FromBinary(string stream_id, byte[] data) =>
        new IpcStreamData(stream_id, data, IPC_DATA_ENCODING.BINARY);
    public static IpcStreamData FromBase64(string stream_id, byte[] data) =>
        new IpcStreamData(stream_id, data.ToBase64(), IPC_DATA_ENCODING.BASE64);
    public static IpcStreamData FromUtf8(string stream_id, byte[] data) =>
        FromUtf8(stream_id, data.ToUtf8());
    public static IpcStreamData FromUtf8(string stream_id, string data) =>
        new IpcStreamData(stream_id, data, IPC_DATA_ENCODING.UTF8);

    private LazyBox<byte[]> _binary = new();
    public byte[] Binary
    {
        get { return _binary.GetOrPut(() => EncodingConverter.DataToBinary(Data, Encoding)); }
    }
    private LazyBox<string> _text = new();
    public string Text
    {
        get { return _text.GetOrPut(() => EncodingConverter.DataToText(Data, Encoding)); }
    }
    public IpcStreamData JsonAble() => Encoding switch
    {
        IPC_DATA_ENCODING.BINARY => FromBase64(
            StreamId,
            (Data as byte[])!
        ),
        _ => this
    };
    /// <summary>
    /// Serialize IpcStreamData
    /// </summary>
    /// <returns>JSON string representation of the IpcStreamData</returns>
    public override string ToJson() => JsonSerializer.Serialize(JsonAble());

    /// <summary>
    /// Deserialize IpcStreamData
    /// </summary>
    /// <param name="json">JSON string representation of IpcStreamData</param>
    /// <returns>An instance of a IpcStreamData object.</returns>
    public static IpcStreamData? FromJson(string json) => JsonSerializer.Deserialize<IpcStreamData>(json);
}

File: android/MicroService/src/main/java/org/dweb_browser/microservice/ipc/helper/IpcStreamData.kt
package org.dweb_browser.microservice.ipc.helper

import com.google.gson.*
import com.google.gson.annotations.JsonAdapter
import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import org.dweb_browser.helper.toBase64
import org.dweb_browser.helper.toUtf8
import java.lang.reflect.Type

@Serializable
data class IpcStreamDataJsonAble(
  val stream_id: String, val encoding: IPC_DATA_ENCODING, val data: String
) : IpcMessage(IPC_MESSAGE_TYPE.STREAM_DATA) {
  fun toIpcStreamData() = IpcStreamData(stream_id, encoding, data)
}

object IpcStreamDataSerializer : KSerializer<IpcStreamData> {
  private val serializer = IpcStreamDataJsonAble.serializer()
  override val descriptor = serializer.descriptor

  override fun deserialize(decoder: Decoder) = serializer.deserialize(decoder).toIpcStreamData()

  override fun serialize(encoder: Encoder, value: IpcStreamData) =
    serializer.serialize(encoder, value.jsonAble)

}

@Serializable(IpcStreamDataSerializer::class)
@JsonAdapter(IpcStreamData::class)
data class IpcStreamData(
  override val stream_id: String,
  val encoding: IPC_DATA_ENCODING,
  val data: Any, /*String or ByteArray*/
) : IpcMessage(IPC_MESSAGE_TYPE.STREAM_DATA), IpcStream, JsonSerializer<IpcStreamData>,
  JsonDeserializer<IpcStreamData> {

  companion object {
    fun fromBinary(streamId: String, data: ByteArray) =
      IpcStreamData(streamId, IPC_DATA_ENCODING.BINARY, data)

    fun fromBase64(streamId: String, data: ByteArray) =
      IpcStreamData(streamId, IPC_DATA_ENCODING.BASE64, data.toBase64())

    fun fromUtf8(streamId: String, data: ByteArray) = fromUtf8(streamId, data.toUtf8())

    fun fromUtf8(streamId: String, data: String) =
      IpcStreamData(streamId, IPC_DATA_ENCODING.UTF8, data)
  }

  val binary by lazy {
    dataToBinary(data, encoding)
  }

  val text by lazy {
    dataToText(data, encoding)
  }

  val jsonAble by lazy {
    when (encoding) {
      IPC_DATA_ENCODING.BINARY -> fromBase64(
        stream_id,
        (data as ByteArray),
      )

      else -> this
    }.run {
      IpcStreamDataJsonAble(stream_id, encoding, data as String)
    }
  }

  override fun serialize(
    src: IpcStreamData, typeOfSrc: Type, context: JsonSerializationContext
  ) = JsonObject().also { jsonObject ->
    with(src.jsonAble) {
      jsonObject.add("type", context.serialize(type))
      jsonObject.addProperty("stream_id", stream_id)
      jsonObject.addProperty("data", data as String)
      jsonObject.add("encoding", context.serialize(encoding))
    }
  }

  override fun deserialize(
    json: JsonElement, typeOfT: Type, context: JsonDeserializationContext
  ): IpcStreamData = json.asJsonObject.let { obj ->
    IpcStreamData(
      stream_id = obj["stream_id"].asString,
      data = obj["data"].asString,
      encoding = context.deserialize(obj["encoding"], IPC_DATA_ENCODING::class.java)
    )
  }
}